### PR TITLE
7 fix create registration back smetanin

### PR DESCRIPTION
**Pull request by dmytrosmetanin**
Changed User model (fix)
created functionality for the registration of the user

### CODE REVIEW

File: codeArena/user/views.py [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
 
-# Create your views here.
+class CreateUserAPIView(APIView):
+    # Allow any user (authenticated or not) to access this url 
+    permission_classes = (AllowAny,)
+    
+
+    # def validate_role_id(self, value):
+    #     if value in ["Moderator", "Admin"]:
+    #         raise ValidationError("U cannot ")
```
Comment (MetalMari) [inline codeArena/user/views.py]: Do you need these commented lines?

File: codeArena/codeArena/urls.py [hunk 2]
```diff
@@ -19,4 +19,5 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('api/task/', include('task.urls')),
+    path('user/', include('user.urls')),
```
Comment (MrSancho42) [inline codeArena/codeArena/urls.py]: We decided to add 'api/' path for API resourses

Comment (MetalMari) [reply codeArena/codeArena/urls.py]: And please, add urls to the user app

File: codeArena/user/models.py [hunk 3]
```diff
@@ -1,28 +1,80 @@
...
+ 
+    def create_user(self, email, password=None, **extra_fields):
+        extra_fields.setdefault('is_staff', False)
+        extra_fields.setdefault('is_superuser', False)
+        return self._create_user(email, password, **extra_fields)
+ 
+    def create_superuser(self, email, password, **extra_fields):
+        extra_fields.setdefault('is_staff', True)
+        extra_fields.setdefault('is_superuser', True)
+
```
Comment (vvelizariy) [inline codeArena/user/models.py]: useless empty line

File: codeArena/user/models.py [hunk 4]
```diff
@@ -1,28 +1,80 @@
...
-    nickname = models.CharField(max_length=70)
+ 
+class UserManager(BaseUserManager):
+ 
+    def _create_user(self, email, password, **extra_fields):
+        """
+        Creates and saves a User with the given email,and password.
+        """
+        if not email:
+            raise ValueError('The given email must be set')
```
Comment (vvelizariy) [inline codeArena/user/models.py:23]: Suppose password is also required.

File: codeArena/user/models.py [hunk 5]
```diff
@@ -1,28 +1,80 @@
...
-    status = models.CharField(max_length=10, choices=[(tag, tag.value) for tag in UserStatus])
+    
+
+    objects = UserManager()
+ 
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name']
+ 
+    def save(self, *args, **kwargs):
+        super(User, self).save(*args, **kwargs)
```
Comment (vvelizariy) [inline codeArena/user/models.py:70]: Will not it work without calling super, just
User.save(*args, **kwargs)
?

File: codeArena/user/models.py [hunk 6]
```diff
@@ -1,28 +1,80 @@
...
+        return self
 
     def __str__(self):
         return f'{self.pk}, {self.email}, {self.nickname}, {self.first_name}, {self.last_name} \
             {(True if self.password is not None else False)}, {self.role_id}, {self.created_at}, {self.updated_at}, {self.status}'
```
Comment (vvelizariy) [inline codeArena/user/models.py]: `
{(True if self.password else False)}
`

File: codeArena/user/serializers.py [hunk 7]
```diff
@@ -6,16 +6,44 @@
...
+
+class UserRegistrationSerializer(serializers.ModelSerializer):
+    password2 = serializers.CharField(
+        write_only=True,
+        required=True
+    )
+
+    class Meta:
+        model = User
+        fields = ['email', 'nickname', 'password', 'password2', 'role_id']
```
Comment (vvelizariy) [inline codeArena/user/serializers.py]: Please name not password2(cause it confuses), but rather verify_password

File: codeArena/user/models.py (hunk 5 above)
Comment (dmytrosmetanin) [reply codeArena/user/models.py:70]: no, it doesn't work
  File "/home/dsprime/CodeArena_project/CodeArena-Django/CodeArena-Django/codeArena/user/models.py", line 70, in save
    User.save(*args, **kwargs)
TypeError: save() missing 1 required positional argument: 'self'

**Review by OleksandrHavron — APPROVED** [review]

File: requirements.txt [hunk 8]
```diff
@@ -1,13 +1,17 @@
...
 Markdown==3.3.4
 psycopg2==2.9.1
+pycodestyle==2.7.0
+PyJWT==1.7.1
 pymongo==3.12.0
 python-decouple==3.4
 pytz==2021.1
 sqlparse==0.2.4
-Pillow==8.3.2
+toml==0.10.2
```
Comment (MrSancho42) [inline requirements.txt:17]: You removed Pillow

File: codeArena/user/urls.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from django.urls import path
+from .views import CreateUserAPIView
+ 
+urlpatterns = [
+    path('create', CreateUserAPIView.as_view()),
+]
```
Comment (MrSancho42) [inline codeArena/user/urls.py:6]: Missing empty row

**Review by MetalMari — APPROVED** [review]

**Review by Maksson — APPROVED** [review]

**Pull request merged**